samples: mesh_badge: fix BT mesh provisioning error

jonas-rem opened the pull request:
The BT Mesh Configuration Client API is used in an asynchronous manner and the app key was not added before attempt binding it to mesh models. This caused the binding functions to fail and resulted in a fail when trying to to send mesh messages.

This commit registers a callback that is called when the app key is assigned. After this callback is called, the binding functions are called.

This PR fixes #54198 

Signed-off-by: Jonas Remmert <j.remmert@phytec.de>

jhedberg reviewed (changes requested):
Thanks for finding the root cause and coming up with a fix proposal! A couple of thoughts:

- Note that printk() doesn’t automatically add a line terminator (new line character). You need to add that yourself. 
- Would it not have been simpler to make the call synchronous, by passing a non-NULL status pointer as the last parameter?

PavelVPV commented:
> * Would it not have been simpler to make the call synchronous, by passing a non-NULL status pointer as the last parameter?

This won't work because underlying layers (transport sar and network loopback) run on the system workqueue. So what you get is a timeout error in the first call and after the timeout the messages will be processed.

PavelVPV commented on samples/boards/reel_board/mesh_badge/src/mesh.c:
```diff
@@ -569,14 +609,20 @@ static void start_mesh(struct k_work *work)
+	}
+}
+
+void show_mesh_status(bool mesh_started_ok, int addr)
```
Could you please move this function up to its forward declaration? We try to avoid to avoid forward declaration unless it is impossible to avoid it.

And could you convert this function to static because it is only used in this file?

PavelVPV commented on samples/boards/reel_board/mesh_badge/src/mesh.c:
```diff
@@ -76,7 +78,67 @@ BT_MESH_HB_CB_DEFINE(hb_cb) = {
+	}
+
+	err = bt_mesh_cfg_cli_mod_pub_set_vnd(NET_IDX, addr, addr, MOD_LF,
+			BT_COMP_ID_LF, &pub, NULL);
```
Looks like this line and some other lines above are not aligned correctly.

jremmert-phytec-iot replied:
fixed

PavelVPV commented on samples/boards/reel_board/mesh_badge/src/mesh.c:
```diff
@@ -76,7 +78,67 @@ BT_MESH_HB_CB_DEFINE(hb_cb) = {
+	/* Bind to vendor model */
+	err = bt_mesh_cfg_cli_mod_app_bind_vnd(NET_IDX, addr, addr, APP_IDX,
+				MOD_LF, BT_COMP_ID_LF, NULL);
+	if (err) {
+		printk("Failed to bind to vendor model (err %d)", err);
+	}
```
Perhaps in case of failure, you could set some boolean variable to false and then pass it to `show_mesh_status`. The rest configuration could be omitted in this case:
```
	bool res = true;

	...

	if (err) {
		printk("Failed to bind to vendor model (err %d)", err);
		res = false;
		goto end;
	}

	...

	printk("Configuration complete\n");
end:
	show_mesh_status(res, addr);
```

PavelVPV reviewed (commented):
Looks good. Since it is a self-configuration, the approach should work because most of the rest configuration messages are unsegmented (except model publication) and also don't depend on each other.

I just have a couple of stylish comments.

Commit:
samples: mesh_badge: fix BT mesh provisioning error

The BT Mesh Configuration Client API is used in an asynchronous manner
and the app key was not added before attempt binding it to mesh models.
This caused the binding functions to fail and resulted in a fail when
trying to to send mesh messages.

This commit registers a callback that is called when the app key is
assigned. After this callback is called, the binding functions are
called.

Signed-off-by: Jonas Remmert <j.remmert@phytec.de>

PavelVPV commented on samples/boards/reel_board/mesh_badge/src/mesh.c:
```diff
@@ -76,7 +76,94 @@ BT_MESH_HB_CB_DEFINE(hb_cb) = {
+				BT_COMP_ID_LF, &pub, NULL);
+	if (err) {
+		printk("Failed to set publication (err %d)\n", err);
+		err = false;
```
No need to change `err`, `show_mesh_status` can be called like this: `show_mesh_status(err == 0, addr)`.

jremmert-phytec-iot replied:
Oh, right, that is easier, will change.

jremmert-phytec-iot commented:
> Thanks for finding the root cause and coming up with a fix proposal! A couple of thoughts:
> 
> * Note that printk() doesn’t automatically add a line terminator (new line character). You need to add that yourself.
> * Would it not have been simpler to make the call synchronous, by passing a non-NULL status pointer as the last parameter?

@PavelVPV elaborated on both variants a) synchronous call from another thread context, b) using the callback,  in the linked issue. I thought b) is easier to understand and opted for it.

Andrewpini reviewed (dismissed):
LGTM 👍

jremmert-phytec-iot commented:
Rebased on current main.
